### ISSUE TITLE
Discussion: distribution preference: file vs. nuget package

### ISSUE THREAD

**Issue by FransBouma** [FransBouma/Massive#223]
This is a discussion topic for the direction in which the code should be going, and especially in the form it is distributed. 

Today all code is in a single file, so distribution is simply copying the file and be done with it. The downside is that all code is copied in all files for each DB, so if a change is needed, all files need to have the change made, there's no code re-use. 

The alternative can be that code is more split up across several files (e.g. a file with code shared among all DBs) and distribution then becomes either copying more files, or using a separate assembly, e.g. distributed through nuget, which would contain the code compiled per DB in a separate assembly (so one for sqlserver, one for oracle etc.). 

Changes are now small and easy to copy, and I haven't looked deeply into the code enough to see whether sharing code is really useful. One thing I'd like to share among the code though is a class which offers string buffers to easily create sql statements (so you can simply append a 'buffer' to a stack of buffers and add elements to that buffer and they all end up at that slot in the stack, and then flatten everything to a string). I have this class already working for many years in the my full ORM LLBLGen Pro. 

Anyway, the TL;DR:
Should:

* code be kept separate per DB in a single file, even if this means copying code in each file and thus making maintaining harder but distribution through code easier

or

* code be shared among classed per DB and distribution then thus is easier through nuget so maintaining the code is easier but distribution is perhaps a bit more tedious. 

(I don't expect a gazillion classes though. Likely 2-3 per DB, so it's not going to be a chore if you then still decide to use file-based distribution of Massive instead of pulling a package from nuget).

**Comment by mountain65** [comment]
Hi,

I like the Nuget approach and the fact that bugfixes easily spread over all the DBs. I have always used Massive as a single separate file in my project and that always felt weird: why is this file from this other guy in my project? But then again, since it was in a separate data-layerish assembly it got pretty soon out of sight.

I think that Rob more-or-less hates NuGet, maybe he can elaborate on that a bit? Preferably in a blog post to enlighten more people on the subject (save keystrokes) ;-)


Richard

**Comment by robconery** [comment]
A bit of history RE: Nuget - the main reason I didn't pop it up there is that the file was never put in the proper directory. Since we don't use a DLL, Nuget was never a good choice.

Originally I intended this to work with WebMatrix so we would drop it into the App_Code directory... doing so, however, marked it as "compile" at build time which causes all kinds of issues.

Anyway - my goal was to allow people to have a "starting point" that they could change as they needed. Take it in any direction you like - but know that one reason I never built out a DLL was because it locked people into the idea that they couldn't change according to their need. 

Massive fits a very unique niche: right below more structured ORMs (like Simple.Data etc).

Just thoughts to chew on :)

**Comment by FransBouma** [comment]
@mountain65 Good points, indeed also what I was thinking. Nuget is a pain, but one can automate it entirely, so it shouldn't be that much of a problem anymore. 

@robconery good point as well. Will have to see if I can find a middle-ground. I don't expect there will be massive amounts of code being added that copying the files will be cumbersome, but having file-based deployment so ppl can alter the code is indeed a good thing to have.

**Comment by panmanphil** [comment]
A variant of the file copying approach but a bit easier on the developer is the new(ish) Shared Project Reference in VS 2013 update 2 and 2015 preview. Developers could just clone the git repo to a folder and include the project file in their solutions. Then the projects that would use Massive just add a project reference to this project. No dlls needed or created. https://visualstudiogallery.msdn.microsoft.com/315c13a7-2787-4f57-bdf7-adae6ed54450
http://channel9.msdn.com/Events/Visual-Studio/Connect-event-2014/517

Definitively a nicer solution when you have multiple framework's to support.

**Comment by FransBouma** [comment]
Phil, that's an awesome idea! haven't thought of that.

**Comment by mountain65** [comment]
Sounds nice,

I haven’t worked with shared projects yet, but this could be a good way to have both the “let massive be in it’s own space” and the easy step-tru-your-code and editting.
Looking forward!


Richard de Zwart

**Comment by FransBouma** [comment]
The shared projects idea is great but relies on an external component still in preview so it's not the best choice at the moment (likely it is in the future, but not now). So that's off the table for now.

**Comment by FransBouma** [comment]
I'm going to try to setup something, it's in the new branch 'Refactoring', to see what works. I'll first try partial classes, and if that doesn't work, inheritance, as a lot of code is sharable among the supported DBs (I guess > 80%). Once this is done, it's much easier to add async/await.

**Comment by FransBouma** [comment]
An interesting idea is shipping code in nuget packages, like https://github.com/mrange/T4Include does

**Comment by panmanphil** [comment]
While it is true that the SharedProject type is a preview only extension, and you are right to be wary of previews, in this case it seems like the risk is very very low compared to the current way of distribution. First it is a part of the VS2015 build so the project type is likely to continue on. Second it is used heavily by Xamarin already. Third, if it doesn't work as it won't with versions of vs 2012 and earlier, the solution reverts to the current one: clone the repository and manually include the file to the projects you want to use it in. 

The T4Include project is a cool idea, and seems to be the same thing as a shared source project. Only as a small github project with 3 contributors, a different set of risks. The nuget distribution is a big plus though

**Comment by ruionwriting** [comment]
For me a pleasant usage would be to have packages like Massive.MySql or Massive.Oracle, etc. So T4Includes looks a good approach.

**Comment by FransBouma** [comment]
Seeing this: http://nikcodes.com/2013/10/23/packaging-source-code-with-nuget/ which defines some conventions, I immediately thought: this isn't for Massive. The whole asp.net native folder layout is something I really don't want to include in the package, but at the same time it's a problem: the code has to go somewhere, and _if_ the user is creating an asp.net application, it's more convenient to add it to that particular folder. 

There is another problem with source-only packages: changing the code they contain is possible, (and one of the reasons they're created in the first place) but it's as risky as changing generated code: if an update of the package is available and the user decides "hey, that's nice, an update, let's pull that into the project", it will overwrite the changes made. 

As I'm a firm believer in that dependencies have to be owned by the person who takes them and it should be as invading as possible (so to make clear to the person taking the dependency what taking a dependency actually _means_) it's best to keep this a source-only github sourced distribution:
- it will be up to the developer to take the dependency and when. 
- when a dependency is taken onto massive, it's part of the project and if an update is released, the developer has to manually make sure that update is mergable into that project. 

I follow @robconery in this, so I'll close this topic now: it will stay as source, distributed here.
